model/relabel/relabel.go

```diff
@@ -192,24 +192,29 @@ func (re Regexp) String() string {
 // are applied in order of input.
 // If a label set is dropped, nil is returned.
 // May return the input labelSet modified.
-func Process(labels labels.Labels, cfgs ...*Config) labels.Labels {
+func Process(lbls labels.Labels, cfgs ...*Config) labels.Labels {
+	lb := labels.NewBuilder(nil)
 	for _, cfg := range cfgs {
-		labels = relabel(labels, cfg)
-		if labels == nil {
+		lbls = relabel(lbls, cfg, lb)
+		if lbls == nil {
 			return nil
 		}
 	}
-	return labels
+	return lbls
 }
 
-func relabel(lset labels.Labels, cfg *Config) labels.Labels {
-	values := make([]string, 0, len(cfg.SourceLabels))
+func relabel(lset labels.Labels, cfg *Config, lb *labels.Builder) labels.Labels {
+	var va [16]string
+	values := va[:0]
+	if len(cfg.SourceLabels) > cap(values) {
+		values = make([]string, 0, len(cfg.SourceLabels))
+	}
```

Comment:
How is this better than the previous?\

Reply:
Also, would re-using the `values` slice between the `relabel` calls make it better?

Reply:
For relabeling rules with up to 16 source labels (surely most real-life examples), it uses an array on the stack, thus performs no heap allocation.

I tried re-using the slice; see https://github.com/prometheus/prometheus/pull/11147/commits/135e5171a5d92dc7b2f5fbff6be133880dd1e949.  It is more complicated and does one allocation in every case, compared to zero in most cases.